README.md
#Practical AES#

PracticalAES currently fully supports EBC and CBC decryption.  A secondary search function that can retrieves decrypted files
based off their file extention (e.x. '.doc', '.sql', '.txt') is partially functional and in the works of being implemented.

###Introduction###

PracticalAES is a function designed to encrypt/decrypt binary files given a cipher key with accordance to AES-128 in both ECB
and CBC operational modes.  It is designed with the intent of simplying the effort requried by the user with regards to not 
only encrypting and decryptiong binary files, but in addition being able to easily lookup and extract files within the binary
code (e.x. .txt, .sql, etc).  

The executable, once compiled, is designed to be easily operated by users.  The EBC mode is enabled by default, but adding an
additional argument to the end automatically enables CBC mode.  The program itself uses no outside libraries, but it uses the
non-standard function itoa().  In the case of your compiler not supporting itoa(), the function can be easily replaced
with a custom, similar itoa() function.

###Usage###

PracticalAES currently supports encryption and decryption for both EBC and CBC modes.  The command to perform said operations
via command line, once compiled, are as followed:

>EBC Encryption: practicalaes.exe e cipherkey binaryfile  
>CBC Encryption: practicalaes.exe e cipherkey binaryfile ivfile  
>EBC Decryption: practicalaes.exe d cipherkey binaryfile  
>CBC Decryption: practicalaes.exe d cipherkey binaryfile ivfile  
>File Retrieval (CBC, IV provided): practicalaes.exe s cipherkey binaryfile .ext iv  
>File Retrieval (CBC, IV assumed to be first 16 bytes): practicalaes.exe s cipherkey binaryfile .ext

Encrypting/Decrypting binary files is fairly straightforward, where the cipher key, binary file to be encrypted/decrypted, as
well as the initialization vector (if CBC is being used) are written in binary files.  For file retrieval, CBC mode is 
the only mode supported and comes with the option of either providing a supplementary argument that provides the iv key
via 16 bytes, or can be performed with an supplementary iv key file, in which it is assumed that the first 16 bytes within
the binary code that is to be decrypted is the initialization vector.

The simplest way create and edit binary files using hex values is via a hex editor.  An example is shown in the image below.

![Cipher](https://cloud.githubusercontent.com/assets/10404525/10374070/f03f1ce6-6da5-11e5-8bf4-ee467acaf66f.PNG)

Because reading in hexadecimal is often the most convinient method when operating with AES, practicalAES was designed with
the intent of being able to read hex files to gather the necessary encryption/decryption keys.

![CmdLine](https://cloud.githubusercontent.com/assets/10404525/10374517/6b6be910-6da8-11e5-933e-668af11308fe.PNG)

A sample executable usage is demonstrated in the command line prompt above.  The code is compiled to an executable, of which
I chose to be practicalaes.exe.  The letter 'e' designates that I would like to perform an encryption, while cipher, sample, 
and iv are all binary files respectively referring to the cipher key, file to be encrypted/decrypted, and the initialization
vector all required for AES.  Note that because the iv file was included at the end, the executable performs a CBC 
encryption.  If the last argument was not provided, the executable would perform a EBC encryption.

It is important to keep in mind that the program uses absolute directories.  This is to be changed by the user, in the code.
The variable for this is (char *basedirectory), and the user is expected to alter this to point to the directory that 
contains both the source code and the binary files required for AES.  The executable will also use the directory to create
an output file.  The default name for the output file is output, although a check is performed to see if output already 
exists, of which then it proceeds to check if output (1), output (2), etc are avaliable to be written in.

The encrypted output for the example demonstrated above can be seen as shown.  Sample contained a total of 48 bytes, and
therefore a 48 byte output is produced.

![Output](https://cloud.githubusercontent.com/assets/10404525/10374518/6beb0768-6da8-11e5-8bbb-699c8519d4d9.PNG)

###Results###

Below are a few tests to demonstrate the accuracy of practicalaes for both the EBC and CBC modes.  Consider the following
inputs.  
```
Binary File: 
32 43 F6 A8 88 5A 30 8D 31 31 98 A2 E0 37 07 34  
89 A7 D9 8E 98 98 7F 97 36 D9 8C FB 38 23 23 98  
27 66 98 C7 9D 87 76 9D 09 34 8D 86 88 78 D1 10  
Cipher Key:  
2B 7E 15 16 28 AE D2 A6 AB F7 15 88 09 CF F4 3C
Initialization Vector:  
43 F4 3D 32 90 19 3E 4B 92 C9 1A 3D 0F FF 2D 7B
```
The resulting outputs for various encryption/decryption scenarios are displayed below.
```
ECB Encryption Result
D9 0D A0 3F BD BD 20 71 E9 37 DB 6E 42 1D D3 62 
6B D9 79 1A 7E 9B 58 A4 84 07 C4 47 12 A5 68 FF  
1C CA 74 05 F5 1B 9E 5B 6B 4E 4C 25 FF 29 89 BD  
ECB Decryption Result
64 F8 C9 20 98 E2 B2 54 E6 09 D2 66 6B CB FD 9A  
41 E1 63 B1 09 C6 F4 B8 B7 43 4B 25 31 18 30 A8  
B0 F7 86 89 B7 18 FB 04 6A CE 04 F3 EF B8 63 7E  
CBC Encryption Result
E8 DE C5 0E 7A D0 26 A8 2C 0D 25 0F 4A 18 C3 BF  
EF 2E 2B 5A 1B 65 07 FA 12 A4 23 67 BA 88 F6 48  

changelog
11/06/2015

Logic for the search functionality has been improved to be more cleaner, accurate, and overall easier to understand.  Usage of
break/continue has been removed and replaced with functions that utilize return values to perform a similar task.

11/01/2015

Implemented basics of search function, which is used to extract individual files from a set of encrypted files depending on
the file extension the user requests.  The general format is explained in the readme.  Currently there exists a few issues
with the function, which will likely be fixed in upcoming versions.

Fixed a slight error regarding the CBC decryption that produced an incorrect result.

The sample EBC/CBC encryption and decryption on the readme were based off a different cipher key and have been corrected 
to demonstrate their correct values.

10/29/2015

The search function is undergoing implementation.  Essentially what occurs is that the user inputs a file extention (i.e. '.sql')
which will then be searched by the post CBC decryption as to whether the file exists or not.  If there exists a file type of 
which the user requests, it will be written out.  The main benefit of this function is that it does not require the user to
save every file that gets decrypted.

The process behind the search function is as follows.  Files are found based off the character '.'.  For example,
'.\Andrew\File1\File2\text.doc' would result in the text.doc folder to be found if I were to be searching for the .doc
extention.  Once the file is found, the contents of the file would be written within their respective folders.

10/08/2015

Fixed minor syntax errors involved with the CBC encryption/decryption.

Changed the cipher key to be read off a binary file as opposed to a text file.

Changed project name to practicalAES.

Updated Readme with descriptions and images of how to use practicalaes.

10/06/2015
EBC and CBC options are fully implemented.  CBC mode is automaticaly enabled by adding an additional char* argument to the end.
Added several functions to make CBC encryption/decryption to essentially be a variation of EBC.  This primarily includes
function involved the variable xorvector, which is exclusively used for the CBC modes.  I am attempting to find a workaround
to help make the code cleaner by only enabling various functions and varibles through the usage of header guards.

The initialization vector is accepted via a byte file of size 16 bytes.  For parity reasons, the cipher, which has originally
been accepted as text format, may be moved to be stored as a byte file.  The primary issue with this is that it complicates

aescrypt.c
#include <stdio.h>
#include <stdlib.h>

#include <stdint.h>
#include <string.h>

#include <unistd.h>
//#include header files


//NOTE THAT FOR 2D ARRAYS, MATRICES ARE COUNTED BY I,J, WHERE I IS THE ROW, AND J IS THE COLUMN!

//typedef uint8_t state_t[4][4];
//static state_t* state;

static uint8_t state[8][8];


static uint8_t roundkey[4][44];

static uint8_t cipherkey[16];



#define Columncount 4
#define Cipherkeylength 16
#define Rounds 10

static const uint8_t sbox[256] =   {
  //0     1    2      3     4    5     6     7      8    9     A      B    C     D     E     F
  0x63, 0x7c, 0x77, 0x7b, 0xf2, 0x6b, 0x6f, 0xc5, 0x30, 0x01, 0x67, 0x2b, 0xfe, 0xd7, 0xab, 0x76,
  0xca, 0x82, 0xc9, 0x7d, 0xfa, 0x59, 0x47, 0xf0, 0xad, 0xd4, 0xa2, 0xaf, 0x9c, 0xa4, 0x72, 0xc0,
  0xb7, 0xfd, 0x93, 0x26, 0x36, 0x3f, 0xf7, 0xcc, 0x34, 0xa5, 0xe5, 0xf1, 0x71, 0xd8, 0x31, 0x15,
  0x04, 0xc7, 0x23, 0xc3, 0x18, 0x96, 0x05, 0x9a, 0x07, 0x12, 0x80, 0xe2, 0xeb, 0x27, 0xb2, 0x75,
  0x09, 0x83, 0x2c, 0x1a, 0x1b, 0x6e, 0x5a, 0xa0, 0x52, 0x3b, 0xd6, 0xb3, 0x29, 0xe3, 0x2f, 0x84,
  0x53, 0xd1, 0x00, 0xed, 0x20, 0xfc, 0xb1, 0x5b, 0x6a, 0xcb, 0xbe, 0x39, 0x4a, 0x4c, 0x58, 0xcf,
  0xd0, 0xef, 0xaa, 0xfb, 0x43, 0x4d, 0x33, 0x85, 0x45, 0xf9, 0x02, 0x7f, 0x50, 0x3c, 0x9f, 0xa8,
  0x51, 0xa3, 0x40, 0x8f, 0x92, 0x9d, 0x38, 0xf5, 0xbc, 0xb6, 0xda, 0x21, 0x10, 0xff, 0xf3, 0xd2,
  0xcd, 0x0c, 0x13, 0xec, 0x5f, 0x97, 0x44, 0x17, 0xc4, 0xa7, 0x7e, 0x3d, 0x64, 0x5d, 0x19, 0x73,
  0x60, 0x81, 0x4f, 0xdc, 0x22, 0x2a, 0x90, 0x88, 0x46, 0xee, 0xb8, 0x14, 0xde, 0x5e, 0x0b, 0xdb,
  0xe0, 0x32, 0x3a, 0x0a, 0x49, 0x06, 0x24, 0x5c, 0xc2, 0xd3, 0xac, 0x62, 0x91, 0x95, 0xe4, 0x79,
  0xe7, 0xc8, 0x37, 0x6d, 0x8d, 0xd5, 0x4e, 0xa9, 0x6c, 0x56, 0xf4, 0xea, 0x65, 0x7a, 0xae, 0x08,
  0xba, 0x78, 0x25, 0x2e, 0x1c, 0xa6, 0xb4, 0xc6, 0xe8, 0xdd, 0x74, 0x1f, 0x4b, 0xbd, 0x8b, 0x8a,
  0x70, 0x3e, 0xb5, 0x66, 0x48, 0x03, 0xf6, 0x0e, 0x61, 0x35, 0x57, 0xb9, 0x86, 0xc1, 0x1d, 0x9e,
  0xe1, 0xf8, 0x98, 0x11, 0x69, 0xd9, 0x8e, 0x94, 0x9b, 0x1e, 0x87, 0xe9, 0xce, 0x55, 0x28, 0xdf,
  0x8c, 0xa1, 0x89, 0x0d, 0xbf, 0xe6, 0x42, 0x68, 0x41, 0x99, 0x2d, 0x0f, 0xb0, 0x54, 0xbb, 0x16 };

static const uint8_t inversesbox[256] =
{ 0x52, 0x09, 0x6a, 0xd5, 0x30, 0x36, 0xa5, 0x38, 0xbf, 0x40, 0xa3, 0x9e, 0x81, 0xf3, 0xd7, 0xfb,
  0x7c, 0xe3, 0x39, 0x82, 0x9b, 0x2f, 0xff, 0x87, 0x34, 0x8e, 0x43, 0x44, 0xc4, 0xde, 0xe9, 0xcb,
  0x54, 0x7b, 0x94, 0x32, 0xa6, 0xc2, 0x23, 0x3d, 0xee, 0x4c, 0x95, 0x0b, 0x42, 0xfa, 0xc3, 0x4e,
  0x08, 0x2e, 0xa1, 0x66, 0x28, 0xd9, 0x24, 0xb2, 0x76, 0x5b, 0xa2, 0x49, 0x6d, 0x8b, 0xd1, 0x25,
  0x72, 0xf8, 0xf6, 0x64, 0x86, 0x68, 0x98, 0x16, 0xd4, 0xa4, 0x5c, 0xcc, 0x5d, 0x65, 0xb6, 0x92,
  0x6c, 0x70, 0x48, 0x50, 0xfd, 0xed, 0xb9, 0xda, 0x5e, 0x15, 0x46, 0x57, 0xa7, 0x8d, 0x9d, 0x84,
  0x90, 0xd8, 0xab, 0x00, 0x8c, 0xbc, 0xd3, 0x0a, 0xf7, 0xe4, 0x58, 0x05, 0xb8, 0xb3, 0x45, 0x06,
  0xd0, 0x2c, 0x1e, 0x8f, 0xca, 0x3f, 0x0f, 0x02, 0xc1, 0xaf, 0xbd, 0x03, 0x01, 0x13, 0x8a, 0x6b,
  0x3a, 0x91, 0x11, 0x41, 0x4f, 0x67, 0xdc, 0xea, 0x97, 0xf2, 0xcf, 0xce, 0xf0, 0xb4, 0xe6, 0x73,
  0x96, 0xac, 0x74, 0x22, 0xe7, 0xad, 0x35, 0x85, 0xe2, 0xf9, 0x37, 0xe8, 0x1c, 0x75, 0xdf, 0x6e,
  0x47, 0xf1, 0x1a, 0x71, 0x1d, 0x29, 0xc5, 0x89, 0x6f, 0xb7, 0x62, 0x0e, 0xaa, 0x18, 0xbe, 0x1b,
  0xfc, 0x56, 0x3e, 0x4b, 0xc6, 0xd2, 0x79, 0x20, 0x9a, 0xdb, 0xc0, 0xfe, 0x78, 0xcd, 0x5a, 0xf4,
  0x1f, 0xdd, 0xa8, 0x33, 0x88, 0x07, 0xc7, 0x31, 0xb1, 0x12, 0x10, 0x59, 0x27, 0x80, 0xec, 0x5f,
  0x60, 0x51, 0x7f, 0xa9, 0x19, 0xb5, 0x4a, 0x0d, 0x2d, 0xe5, 0x7a, 0x9f, 0x93, 0xc9, 0x9c, 0xef,
  0xa0, 0xe0, 0x3b, 0x4d, 0xae, 0x2a, 0xf5, 0xb0, 0xc8, 0xeb, 0xbb, 0x3c, 0x83, 0x53, 0x99, 0x61,
  0x17, 0x2b, 0x04, 0x7e, 0xba, 0x77, 0xd6, 0x26, 0xe1, 0x69, 0x14, 0x63, 0x55, 0x21, 0x0c, 0x7d };

static const uint8_t Rcon[256] = {
  0x8d, 0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80, 0x1b, 0x36, 0x6c, 0xd8, 0xab, 0x4d, 0x9a,
  0x2f, 0x5e, 0xbc, 0x63, 0xc6, 0x97, 0x35, 0x6a, 0xd4, 0xb3, 0x7d, 0xfa, 0xef, 0xc5, 0x91, 0x39,
  0x72, 0xe4, 0xd3, 0xbd, 0x61, 0xc2, 0x9f, 0x25, 0x4a, 0x94, 0x33, 0x66, 0xcc, 0x83, 0x1d, 0x3a,
  0x74, 0xe8, 0xcb, 0x8d, 0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80, 0x1b, 0x36, 0x6c, 0xd8,
  0xab, 0x4d, 0x9a, 0x2f, 0x5e, 0xbc, 0x63, 0xc6, 0x97, 0x35, 0x6a, 0xd4, 0xb3, 0x7d, 0xfa, 0xef,
  0xc5, 0x91, 0x39, 0x72, 0xe4, 0xd3, 0xbd, 0x61, 0xc2, 0x9f, 0x25, 0x4a, 0x94, 0x33, 0x66, 0xcc,
  0x83, 0x1d, 0x3a, 0x74, 0xe8, 0xcb, 0x8d, 0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80, 0x1b,
  0x36, 0x6c, 0xd8, 0xab, 0x4d, 0x9a, 0x2f, 0x5e, 0xbc, 0x63, 0xc6, 0x97, 0x35, 0x6a, 0xd4, 0xb3,
  0x7d, 0xfa, 0xef, 0xc5, 0x91, 0x39, 0x72, 0xe4, 0xd3, 0xbd, 0x61, 0xc2, 0x9f, 0x25, 0x4a, 0x94,
  0x33, 0x66, 0xcc, 0x83, 0x1d, 0x3a, 0x74, 0xe8, 0xcb, 0x8d, 0x01, 0x02, 0x04, 0x08, 0x10, 0x20,
  0x40, 0x80, 0x1b, 0x36, 0x6c, 0xd8, 0xab, 0x4d, 0x9a, 0x2f, 0x5e, 0xbc, 0x63, 0xc6, 0x97, 0x35,
  0x6a, 0xd4, 0xb3, 0x7d, 0xfa, 0xef, 0xc5, 0x91, 0x39, 0x72, 0xe4, 0xd3, 0xbd, 0x61, 0xc2, 0x9f,
  0x25, 0x4a, 0x94, 0x33, 0x66, 0xcc, 0x83, 0x1d, 0x3a, 0x74, 0xe8, 0xcb, 0x8d, 0x01, 0x02, 0x04,
  0x08, 0x10, 0x20, 0x40, 0x80, 0x1b, 0x36, 0x6c, 0xd8, 0xab, 0x4d, 0x9a, 0x2f, 0x5e, 0xbc, 0x63,
  0xc6, 0x97, 0x35, 0x6a, 0xd4, 0xb3, 0x7d, 0xfa, 0xef, 0xc5, 0x91, 0x39, 0x72, 0xe4, 0xd3, 0xbd,
  0x61, 0xc2, 0x9f, 0x25, 0x4a, 0x94, 0x33, 0x66, 0xcc, 0x83, 0x1d, 0x3a, 0x74, 0xe8, 0xcb, 0x8d };

static uint8_t SBoxSub (uint8_t hex){
    return sbox[hex];
}

static uint8_t ISBoxSub (uint8_t hex){
    return inversesbox[hex];
}

static void RoundKeyExpansion (void){

//First 16 bytes of the round key is a copy of the cipher key
    int i, j;
    uint8_t placeholder[4];
    for (i = 0; i < Columncount; i++){
        roundkey[0][i] = cipherkey[4*i + 0];
        roundkey[1][i] = cipherkey[4*i + 1];
        roundkey[2][i] = cipherkey[4*i + 2];
        roundkey[3][i] = cipherkey[4*i + 3];
    }

//Take last existing column, solve for the remainder of the round keys
//Note that first column is a swap, substitution with Rcon, than add
//The remainder of the columns are addition to Rcon

//Aggregate of one, because these are to count from 1 to 10, as 0 was solved above
    for (i = 4; i < (Rounds + 1) * Columncount; i++){

        //Now if the column count is divisible by 4, then you rotate the last letter to the start
        //column count is equal to i

        //Afterwards, you then perform a s box substitution
        if (i % 4 == 0){
            for (j = 0; j < 4; j++){
                placeholder[j] = roundkey[j][i-1];
            }
            uint8_t tempholder;
            tempholder = placeholder[0];
            placeholder[0] = placeholder[1];
            placeholder[1] = placeholder[2];
            placeholder[2] = placeholder[3];
            placeholder[3] = tempholder;

            placeholder[0] = SBoxSub(placeholder[0]);
            placeholder[1] = SBoxSub(placeholder[1]);
            placeholder[2] = SBoxSub(placeholder[2]);
            placeholder[3] = SBoxSub(placeholder[3]);

            //Rcon [1x4] is then added to all columns that are of multiple of 4 (AKA start of each round key)
            //Note that only the first index of Rcon is non zero, and that 0x8d (Rcon[0]) is skipped!
            //Round Key 1 would include 0x01 to the first hex value

            placeholder[0] = placeholder[0] ^ Rcon[i/4];

            roundkey[0][i] = placeholder[0] ^ roundkey[0][i - 4];
            roundkey[1][i] = placeholder[1] ^ roundkey[1][i - 4];
            roundkey[2][i] = placeholder[2] ^ roundkey[2][i - 4];
            roundkey[3][i] = placeholder[3] ^ roundkey[3][i - 4];
        } else {
            roundkey[0][i] = roundkey[0][i-4] ^ roundkey[0][i - 1];
            roundkey[1][i] = roundkey[1][i-4] ^ roundkey[1][i - 1];
            roundkey[2][i] = roundkey[2][i-4] ^ roundkey[2][i - 1];
            roundkey[3][i] = roundkey[3][i-4] ^ roundkey[3][i - 1];
        }
    }


}

//Rounds 1 - 10
//SubBytes
//Shift Rows
//Mix Columns
//Add Round Key

//Round 11
//SubBytes
//Shift Rows
//Add Round Key

static void ShiftRows(void){
    //shift the rows in the 16 byte holder state
    //First row, no rotations
    //Second row, rotate over by 1 byte (1 2 3 4 becomes 2 3 4 1)
    //Third row, rotate over by 2 bytes (1 2 3 4 becomes 3 4 1 2 )
    //Fourth row, rotate over by 3 bytes (1 2 3 4 becomes 4 1 2 3)

    uint8_t rowholder[1][3];

    rowholder[0][0] = state[1][0];
    state[1][0] = state[1][1];
    state[1][1] = state[1][2];
    state[1][2] = state[1][3];
    state[1][3] = rowholder[0][0];

    rowholder[0][0] = state[2][0];
    rowholder[0][1] = state[2][1];
    state[2][0] = state[2][2];
    state[2][1] = state[2][3];
    state[2][2] = rowholder[0][0];
    state[2][3] = rowholder[0][1];

    rowholder[0][0] = state[3][0];
    rowholder[0][1] = state[3][1];
    rowholder[0][2] = state[3][2];
    state[3][0] = state[3][3];
    state[3][1] = rowholder[0][0];
    state[3][2] = rowholder[0][1];
    state[3][3] = rowholder[0][2];

}

static void InverseShiftRows(void){
    //ShiftRows, but backwards
    uint8_t inverserowholder[1][3];

    //Shift second row all elements right once, rotate over
    inverserowholder[0][0] = state[1][3];
    state[1][3] = state[1][2];
    state[1][2] = state[1][1];
    state[1][1] = state[1][0];
    state[1][0] = inverserowholder[0][0];

    //Shift third row to the right twice, rotate over
    inverserowholder[0][0] = state[2][3];
    inverserowholder[0][1] = state[2][2];
    state[2][3] = state[2][1];
    state[2][2] = state[2][0];
    state[2][1] = inverserowholder[0][0];
    state[2][0] = inverserowholder[0][1];

    //Shift fourth row to the right three times, rotate over
    inverserowholder[0][0] = state[3][3];
    inverserowholder[0][1] = state[3][2];
    inverserowholder[0][2] = state[3][1];
    state[3][3] = state[3][0];
    state[3][2] = inverserowholder[0][0];
    state[3][1] = inverserowholder[0][1];
    state[3][0] = inverserowholder[0][2];
}

//MixColumns multiplies a column of 4 bytes by a 4x4 matrix of the form
//{ 02 03 01 01
//  01 02 03 01
//  01 01 02 03
//  03 01 01 02 }

//see https://en.wikipedia.org/wiki/Rijndael_mix_columns for explanation
//The concept here is that you would want to do 2 + 2 + 1 + 1 + 1
//where you solely use multipliers of 1 or 2, creating 2 by using a bitshift.
static void MixColumns(void){
    uint8_t singlestate[4][4];
    uint8_t doublestate[4][4];
    uint8_t tempholder;
    int i, j;
    for (i = 0; i < 4; i++){
        for (j = 0; j < 4; j++){
            singlestate[j][i] = state[j][i];
            tempholder = (unsigned char)((signed char)state[j][i] >> 7);
            doublestate[j][i] = state[j][i] << 1;
            doublestate[j][i] ^= 0x1B & tempholder;
        }
        state[0][i] = doublestate[0][i] ^ singlestate[3][i] ^ singlestate[2][i] ^ doublestate[1][i] ^ singlestate[1][i];
        state[1][i] = doublestate[1][i] ^ singlestate[0][i] ^ singlestate[3][i] ^ doublestate[2][i] ^ singlestate[2][i];
        state[2][i] = doublestate[2][i] ^ singlestate[1][i] ^ singlestate[0][i] ^ doublestate[3][i] ^ singlestate[3][i];
        state[3][i] = doublestate[3][i] ^ singlestate[2][i] ^ singlestate[1][i] ^ doublestate[0][i] ^ singlestate[0][i];

    }
}

//InverseMixColumns multiplies a column of 4 bytes by a 4x4 matrix of the form
//{ 14 11 13 09
//  09 14 11 13
//  13 09 14 11
//  11 13 09 14 }

//Method referenced from https://github.com/kokke/tiny-AES128-C/blob/master/aes.c
//To do this, a function Multiply is defined using bit shifts
//This may be edited out later for another method,

//http://crypto.stackexchange.com/questions/2569/how-does-one-implement-the-inverse-of-aes-mixcolumns
static uint8_t Multiply(uint8_t x, int y){
    uint8_t returnvalue, highbit;

    if (y == 9){
        // (((x * 2) * 2) * 2) + x
        highbit = (unsigned char)((signed char)x >> 7);
        returnvalue = x << 1;
        returnvalue  ^= 0x1B & highbit;

        highbit = (unsigned char)((signed char)returnvalue >> 7);
        returnvalue = returnvalue << 1;
        returnvalue ^= 0x1B & highbit;

        highbit = (unsigned char)((signed char)returnvalue >> 7);
        returnvalue = returnvalue << 1;
        returnvalue ^= 0x1B & highbit;

        returnvalue ^= x;


    } else if (y == 11) {
        //((((x * 2) * 2) + x) * 2) + x
        highbit = (unsigned char)((signed char)x >> 7);
        returnvalue = x << 1;
        returnvalue  ^= 0x1B & highbit;

        highbit = (unsigned char)((signed char)returnvalue >> 7);
        returnvalue = returnvalue << 1;
        returnvalue ^= 0x1B & highbit;

        returnvalue ^= x;

        highbit = (unsigned char)((signed char)returnvalue >> 7);
        returnvalue = returnvalue << 1;
        returnvalue ^= 0x1B & highbit;

        returnvalue ^= x;

    } else if (y == 13) {
        //((((x * 2) + x) * 2) * 2) + x
        highbit = (unsigned char)((signed char)x >> 7);
        returnvalue = x << 1;
        returnvalue  ^= 0x1B & highbit;

        returnvalue ^= x;

        highbit = (unsigned char)((signed char)returnvalue >> 7);
        returnvalue = returnvalue << 1;
        returnvalue ^= 0x1B & highbit;

        highbit = (unsigned char)((signed char)returnvalue >> 7);
        returnvalue = returnvalue << 1;
        returnvalue ^= 0x1B & highbit;

        returnvalue ^= x;

    } else if (y == 14) {
        //((((x * 2) + x) * 2) + x) * 2
        highbit = (unsigned char)((signed char)x >> 7);
        returnvalue = x << 1;
        returnvalue  ^= 0x1B & highbit;

        returnvalue ^= x;

        highbit = (unsigned char)((signed char)returnvalue >> 7);
        returnvalue = returnvalue << 1;
        returnvalue ^= 0x1B & highbit;

        returnvalue ^= x;

        highbit = (unsigned char)((signed char)returnvalue >> 7);
        returnvalue = returnvalue << 1;
        returnvalue ^= 0x1B & highbit;

    } else {
        returnvalue = 0x00;
    }

    return returnvalue;
}


static void InverseMixColumns(void){
    int i,j;
    uint8_t tempholder[4];
    for ( i = 0; i < 4; i++){
        for ( j = 0; j < 4; j++){
            tempholder[j] = state[j][i];

        }
        state[0][i] = Multiply(tempholder[0], 14) ^ Multiply(tempholder[1], 11) ^ Multiply(tempholder[2], 13) ^ Multiply(tempholder[3], 9);
        state[1][i] = Multiply(tempholder[0], 9) ^ Multiply(tempholder[1], 14) ^ Multiply(tempholder[2], 11) ^ Multiply(tempholder[3], 13);
        state[2][i] = Multiply(tempholder[0], 13) ^ Multiply(tempholder[1], 9) ^ Multiply(tempholder[2], 14) ^ Multiply(tempholder[3], 11);
        state[3][i] = Multiply(tempholder[0], 11) ^ Multiply(tempholder[1], 13) ^ Multiply(tempholder[2], 9) ^ Multiply(tempholder[3], 14);
    }
}

static void AddRoundKey(int round){
    int i, j;
    for (i = 0; i < 4; i++){
        for ( j = 0; j < 4; j++){
            state[j][i] ^= roundkey[j][i + 4*round];
        }
    }

}

static void SubstituteBytes(void){
    int i, j;
    for (i = 0; i < 4; i++){
        for (j = 0; j < 4; j++){
            (state)[i][j]  = SBoxSub((state)[i][j]);
        }
    }
}

static void InverseSubstituteBytes(void){
    int i, j;
    for (i = 0; i < 4; i++){
        for (j = 0; j < 4; j++){
            (state)[i][j]  = ISBoxSub((state)[i][j]);
        }
    }
}

static void PrintState(void){
    int i,j;
    printf("\n");
    for (i = 0; i < 4; i++){
        for (j = 0; j < 4; j++){
            printf("%02X", state[j][i]);
        }
    }
}

static void EncryptRounds(void){
    int i;
    PrintState();
    RoundKeyExpansion();
    AddRoundKey(0);
    for (i = 1; i < Rounds; i++){
        SubstituteBytes();
        ShiftRows();
        MixColumns();
        AddRoundKey(i);

    }
    SubstituteBytes();
    ShiftRows();
    AddRoundKey(i);
}


static void DecryptRounds(void){
    int i;
    RoundKeyExpansion();
    AddRoundKey(10);
    InverseShiftRows();
    InverseSubstituteBytes();

    for (i = 9; i > 0; i--){
        AddRoundKey(i);
        InverseMixColumns();
        InverseShiftRows();
        InverseSubstituteBytes();
    }

    AddRoundKey(i);
}

int converthexvalue(uint8_t h){
    int value = (int)h;
    if(value < 58 && value > 47){
        return value - 48;
    }
    if(value < 103 && value > 96){
        return value- 87;
    }
    return value;
}

static void RetrieveCipherKey(char* cipherdirectory){
    FILE *keydata = fopen(cipherdirectory, "rt");
    if (keydata == NULL){
        perror("Error");
        printf("Error occurred while retrieving cipher key.");
    }
    uint8_t leftvalue, rightvalue, hexvalue;
    int i;
    for (i = 0; i < 16; i++){
        leftvalue = converthexvalue(fgetc(keydata));
        rightvalue = converthexvalue(fgetc(keydata));
        hexvalue = leftvalue << 4 | rightvalue;
        cipherkey[i] = hexvalue;
    }
    fclose(keydata);
}

static char* SearchOutputFileName(char* outputfilename){
    int outputexists = 1;
    int iteration = 0;
    while(outputexists == 1){
        if (access(outputfilename, F_OK) != -1){
            //create new outputfilename and loop again
            char *strbase = "output (";
            char *strend = ")";
            char outputfilestr[20];
            char iterationvalue[10];

            iteration++;

            strcpy(outputfilestr, "output (");
            itoa(iteration,iterationvalue, 10);
            //strcpy(iterationvalue, iteration);
            strcat(outputfilestr, iterationvalue);
            //outputfilestr = strcat(strbase, iteration);
            strcat(outputfilestr, strend);
            outputfilename = malloc(1 + strlen(outputfilestr));
            strcpy(outputfilename, outputfilestr);
            outputfilename[1 + strlen(outputfilestr)] = '\0';

            free(strbase);
            free(strend);
        } else {
            outputexists = 0;
        }
    }
    return outputfilename;
}


//d - decrypt (exe, d, cipher, input file) - outputs decrypted file EBC
//d - decrypt (exe, d, cipher, input file, iv) - outputs decrypted file CBC
//e - encrypt (exe, d, cipher, input file) - outputs encrypted file EBC
//e - encrypt (exe, e, cipher, input file, iv) - outputs encrypted file CBC

//s - specific directory retrieval EBC(exe, s, cipher, inputfile, filetype)
//s - specific directory retrieval CBC(exe, s, cipher, inputfile, filetype, iv)



int main(int argc, char *argv[]){
    //Retrieve Key
    char *cipherdirectory;
    char *targetdirectory;
    char *basedirectory = "C:\\Users\\Andrew\\cworkspace\\AESdecrypt\\";

    if((cipherdirectory = malloc(strlen(basedirectory)+strlen(argv[2])+1)) != NULL){
        cipherdirectory[0] = '\0';   // ensures the memory is an empty string
        strcat(cipherdirectory, basedirectory);
        strcat(cipherdirectory, argv[2]);
    } else {
        perror("Error");
        printf("Error occurred with allocating memory while searching for cipher key.");
        return 0;
    }

    RetrieveCipherKey(cipherdirectory);

    //Checks to see if input and output files are valid
    if((targetdirectory = malloc(strlen(basedirectory)+strlen(argv[3])+1)) != NULL){
        targetdirectory[0] = '\0';   // ensures the memory is an empty string
        strcat(targetdirectory, basedirectory);
        strcat(targetdirectory, argv[3]);
    } else {
        perror("Error");
        printf("Error occurred with allocating memory while searching for binary file.");
    }

    FILE *inputfile, *outputfile;
    inputfile = fopen(targetdirectory, "rb");
    char *outputfilename = "output";
    outputfilename = SearchOutputFileName(outputfilename);
    outputfile = fopen(outputfilename, "ab+");

    if (inputfile == NULL || outputfile == NULL){
        printf("Error occurred while opening input or output files.\n");
        return 0;
    }

    int counted = 0;
    int read = 0;
    uint8_t buffer[16];
    int i;

    if (strcmp(argv[1],"d") == 0){
        if (argc == 4){
            //EBC decrypt
            while((read = fread(buffer, 1, 16, inputfile)) > 0){
                counted = counted + read;
                for(i=0; i < 16; i++){
                    if (i < read){
                        state[i%4][(int)i/4] = buffer[i];
                    } else {
                        state[i%4][(int)i/4] = (uint8_t)0x00;
                    }
                }
                DecryptRounds();
                for(i=0; i < 16; i++){
                    fwrite(&state[i%4][(int)i/4],sizeof(uint8_t), 1, outputfile);
                }
            }
        } else if (argc == 5) {
            //CBC decrypt
        } else {
            //Error
            printf("Input format for decryption is invalid.\n");
        }
    } else if (strcmp(argv[1],"e") == 0) {
        if (argc == 4){
            //EBC encrypt
            while((read = fread(buffer, 1, 16, inputfile)) > 0){
                counted = counted + read;
                for(i=0; i < 16; i++){
                    if (i < read){
                        state[i%4][(int)i/4] = buffer[i];
                    } else {
                        state[i%4][(int)i/4] = (uint8_t)0x00;
                    }
                }
                EncryptRounds();

                for(i=0; i < 16; i++){
                    fwrite(&state[i%4][(int)i/4],sizeof(uint8_t), 1, outputfile);
                }
            }
        } else if (argc == 5){
            //CBC encrypt
        } else {
            //Error
            printf("Input format for encryption is invalid.\n");
        }
    } else if (strcmp(argv[1],"s") == 0) {
        //Currently in progress
        printf("Function is currently not supported.\n");
    } else {
        printf("Invalid input format.\n");
    }

    fclose(inputfile);
    fclose(outputfile);
    //if ivfile is open, close it
    return 0;
}
